### DISCUSSION TITLE
A header access API?

### DISCUSSION THREAD

**Discussion by bagder** [curl/curl#8335] [Ideas]
I'm proposing a new API for accessing and extracting (HTTP) header contents after a transfer, and I'm looking for feedback to help us making it really useful. It is especially helpful if you have an application or specific use cases in use or in mind and we make sure those use cases can be satisfied by the API.

The API proposal as just now: https://github.com/curl/curl/wiki/get-headers

(again: these are ideas and thoughts, not a final concept)

**Comment by jzakrzewski** [comment]
I'd love to see an API that gives the headers correctly - takes care about (un-)escaping, splitting in case of multiple values, etc.
From the API perspective there should be no difference between multiple headers of the same name and a header with multiple, comma-separated values. 
Only the cookies concern me because whoever conceived  syntax for those headers did it in complete disregard to what the comma means for other headers...

At best I'd like to:
1. Get a comma-separated string when I simply ask for header value
2. Be able to ask if it's multi-value header and then have access to each value separately
3. Somehow get the cookie headers right (maybe just forbid asking for single value for those headers?)

**Reply by bagder** [reply]
I don't believe in merging multiple ones into a single comma-separated value, even if this in theory would be what HTTP says we can do. Because:
1. I don't trust all users to work that out nor to servers to always provide headers to work like that
2. cookies can't work like that
3. not even curl can work out all headers comma-separated but need them separated, like `WWW-Authenticate`

I would rather have the API signal something like "there's another header with the same name and ..."

**Reply by jzakrzewski** [reply]
I understand. From the other hand, applications have to deal with that, which means there's a lot of code doing "the same thing" written by people with various understanding of what they are doing. I don't even want to know how many bugs is in my implementation currently.

Having a single implementation would benefit everyone IMHO.
I don't mind not having the merging thing. After all I can easily do this on a case-by-case basis. Correct splitting would, however, be a godsend.

**Comment by monnerat** [comment]
If extraction means generic parsing (i.e.: token decomposition), I think it's evil: although there are general rules for the syntax, every single header is a particular case. Considering combinations of:
- parameters syntax variants,
- quoting or not,
- escaping methods,
- i18n,
- folded headers,
- space interpretation, delimiters,
- embedded comments,
- and more I forget just now,

this seems an impossible mission.

I attempted something like that a few years ago and finally gave up.

As an example of craziness, remember https://github.com/curl/curl/commit/b20b364764cca2d577640126494fe2b09ae22e5d: it was about mime/form, but headers are even worse. In addition, the task is to parse them, which is much harder than generation.

Very ambitious project, but I fear it's a Pandora box. That said, I'm not against it!

**Reply by bagder** [reply]
The current proposal does not include most of those details...

**Reply by monnerat** [reply]
Yes, I've read it, but as you see in first comment above, there's already a request for more...

**Reply by jzakrzewski** [reply]
I was just hoping for a solid (at least partial) implementation of header decomposition. I had to do such generic things in the past and it's a pain in the back side. I had to limit myself to just looking for commas and give users an escape-hatch if something crazy comes up...

**Reply by monnerat** [reply]
> it's a pain in the back side

I couldn't say it better ! You just confirm my comment.

**Reply by jzakrzewski** [reply]
For me it's a reason to have it implemented by competent people, so that the implementation could be re-used.

**Comment by michael-o** [comment]
The API is incompete, sone headers are allowed to appear multiple times in the response, I don't see how the new API with a mere `char *` will cover this.

**Reply by bagder** [reply]
Hence:

> Question: how about responses with multiple headers using the same name ? Bad ones like Content-Length or accepted ones like cookies

**Reply by michael-o** [reply]
Noticed too late. I wonder, since curl provides an API to deliver request headers with a linked list, can this structure be reused for responses as well?

**Comment by skgsergio** [comment]
It's been long since I coded against curl directly, however, this is a feature I missed. Overall looks nice and fits what I remember from the API.

However regarding the questions, here are my 2cents:

> Question: how about repsonses with multiple headers using the same name ? Bad ones like Content-Length or accepted ones like cookies.

This is a bit of a minefield, is curl's duty to understand which headers can be specified more than once and which ones? This means keeping up with all RFCs (which already does for protocols, but I feel this different)

> Question: should we differ between "there was no header using this name in the response" and "there was no response at all so no headers were returned" in the return? Or how about "there were two headers using this name in the response".

I don't know/remember how other API functions handle the there was no response at all (was it checking if perform returned `!= CURL_OK` or something like that? or there is a case that returns OK but there is no response at all?), but I guess that should be done in the same way.

Maybe the "natural" way I would expect is a list of values (but I might be influenced by other libs in other languages) but I'm not sure if this fits well with the rest of the API.
If this was not an option, maybe an extra function to check how many occurrences there are and in the `curl_easy_header` add an "index" to fetch the different values. Even if 80% of the time we are getting index 0.

> Question: leading and trailing whitespace trim? I think leave in as delivered.

100% agree. If curl trims it automatically it might cause issues if someone is doing something "strange" with it. If the content is "`foobar    `" I think it should be returned as delivered.

> Question: what if the response contains two headers of the same name, how should the application get the second value?

If the issue with multiple values is solved by either returning a list or having a way of checking occurrences and fetching them by "index" then `curl_easy_getheaders` can just return unique occurrences.

**Reply by telc** [reply]
Getting the header by index is probably easiest. Having a prototype without index that just gets index 0 as well.
Finding the number could either be having multiples returned by `curl_easy_getheaders()` or by having something like
`int curl_easy_header_count(CURL *easy, const char *name);`

**Comment by telc** [comment]
> Question: should this require being enabled in the request or should it always do it?

Depends on how much having it always on impacts performance. If it's not negligible, I would prefer a `CURLOPT_EXTRACT_HEADERS` or similar.

**Comment by eiszapfen2000** [comment]
How about
`struct curl_slist *curl_easy_getheaders(CURL *easy);`
like suggested by @bagder here: https://github.com/curl/curl/wiki/get-headers#list-all-headers

But make it a list of key-value pairs of the headers, as received. That way duplicate entries won't be a problem.

`curl_slist->data` is the name of the first header, `curl_slist->next->data` is the actual value of the first header. ASOASF.

**Comment by bagder** [comment]
Time for version 2. I create a [new thread](https://github.com/curl/curl/discussions/8496) for that.
